multicluster/controllers/multicluster/label_identity_controller.go

```diff
@@ -154,160 +146,79 @@ func (r *LabelIdentityReconciler) namespaceMapFunc(ns client.Object) []reconcile
 	return requests
 }
 
-// onPodDelete removes the Pod and label identity mapping from the cache, and
-// updates label identity ResourceExport if necessary (the Pod deletion event
-// causes a label identity no longer present in the cluster).
-func (r *LabelIdentityReconciler) onPodDelete(ctx context.Context, podNamespacedName string) error {
-	labelToDelete := r.getLabelToDelete(podNamespacedName)
-	if labelToDelete != "" {
-		if err := r.deleteLabelIdentityResExport(ctx, labelToDelete); err != nil {
-			return err
-		}
-		r.deleteFromLabelCache(podNamespacedName, labelToDelete)
+// onPodCreateOrUpdate handles Pod create or update events.
+func (r *LabelIdentityReconciler) onPodCreateOrUpdate(ctx context.Context, podNamespacedName, currentLabel string) error {
+	if err := r.checkForLabelResExportDelete(ctx, podNamespacedName, currentLabel); err != nil {
+		return err
 	}
-	return nil
+	return r.checkForLabelResExportAdd(ctx, podNamespacedName, currentLabel)
 }
 
-// onPodCreateOrUpdate updates the Pod and label identity mapping in the cache, and
-// updates label identity ResourceExport if necessary (the Pod creation/update
-// event causes a new label identity to appear in the cluster or a label identity
-// no longer present in the cluster).
-func (r *LabelIdentityReconciler) onPodCreateOrUpdate(ctx context.Context, podNamespacedName, normalizedLabel string) error {
-	labelToAdd, labelToDelete := r.getLabelsToUpdate(podNamespacedName, normalizedLabel)
-	if labelToDelete != "" {
-		if err := r.deleteLabelIdentityResExport(ctx, labelToDelete); err != nil {
-			return err
-		}
-		r.deleteFromLabelCache(podNamespacedName, labelToDelete)
-	}
-	if labelToAdd != "" {
-		if err := r.createLabelIdentityResExport(ctx, labelToAdd); err != nil {
-			return err
+// checkForLabelResExportDelete removes the Pod and label identity mapping from the cache,
+// and deletes label identity ResourceExport if necessary (the Pod deletion event
+// causes a label identity no longer present in the cluster).
+func (r *LabelIdentityReconciler) checkForLabelResExportDelete(ctx context.Context, podNamespacedName, currentLabel string) error {
+	originalLabel, isCached := r.podLabelCache[podNamespacedName]
+	if isCached && originalLabel != currentLabel {
+		unlock := r.labelMutexes.Lock(originalLabel)
+		defer unlock()
```

Comment:
There is another question to think about - when you remove the label from labelMutexes? You may need a reference count? Could you think about that? Earlier, I did not think about a separate map for mutexes. Not sure.

Reply:
I assumed that when you asked what is the overhead of a mutex, you meant whether it is fine to keep a map like this? pruning un-referenced label mutex can be really hard.....

Reply:
Also to answer the mem overhead question, a Mutex is a simple struct with two int32 fields, so 8 bytes each.

Reply:
Mutex is small but acquring it takes time. The method acquires lock 3 times and operate sync.Map 2 twice, each of which acquires a lock once, so it acquires locks 5 times. I doubt it really performs better than 1 mutex with normal maps. Have you verified it?

Reply:
No, I think we should still remove unused mutexes. Cannot keep them for ever.

Reply:
@tnqn @jianjuns Please check the latest change as after a discussion with Jianjun we decided to stick with the re-queue approach to make the logic more straightforward. There's now a single mutex and Pod deletion events now acquires mutex twice max and Pod update events acquires mutex 3 times max.